File: src/commands/link.rs
//! Commands for linking and managing idea relationships

use crate::database_simple as database;
use crate::errors::Result;
use colored::*;
use dialoguer::Confirm;

#[derive(clap::Subcommand, Debug)]
pub enum LinkCommands {
    /// Link two ideas together with a relationship
    Create {
        /// Source idea ID to link from
        #[arg(required = true)]
        source_id: String,

        /// Target idea ID to link to
        #[arg(required = true)]
        target_id: String,

        /// Type of relationship
        #[arg(required=true, value_parser = ["depends_on", "related_to", "part_of", "parent", "child", "duplicate", "blocks", "blocked_by", "similar_to"])]
        relationship_type: String,

        /// Confirm before creating the relationship
        #[arg(short, long)]
        confirm: bool,
    },

    /// List all relationships for an idea
    List {
        /// Idea ID to show relationships for
        #[arg(required = true)]
        idea_id: String,
    },

    /// Show related ideas for an idea
    Show {
        /// Idea ID to show related ideas for
        #[arg(required = true)]
        idea_id: String,

        /// Filter by relationship type
        #[arg(long)]
        relationship_type: Option<String>,
    },

    /// Remove a relationship between ideas
    Remove {
        /// Relationship ID to remove
        #[arg(required = true)]
        relationship_id: String,

        /// Confirm before removing
        #[arg(short, long)]
        confirm: bool,
    },

    /// Find dependency paths between ideas
    Path {
        /// Starting idea ID
        #[arg(required = true)]
        from: String,

        /// Target idea ID
        #[arg(required = true)]
        to: String,
    },
}

pub async fn handle_link(cmd: LinkCommands, db: &database::Database) -> Result<()> {
    match cmd {
        LinkCommands::Create {
            source_id,
            target_id,
            relationship_type,
            confirm,
        } => handle_create_link(source_id, target_id, relationship_type, confirm, db).await,
        LinkCommands::List { idea_id } => handle_list_links(idea_id, db).await,
        LinkCommands::Show {
            idea_id,
            relationship_type,
        } => handle_show_related(idea_id, relationship_type, db).await,
        LinkCommands::Remove {
            relationship_id,
            confirm,
        } => handle_remove_link(relationship_id, confirm, db).await,
        LinkCommands::Path { from, to } => handle_find_path(from, to, db).await,
    }
}

async fn handle_create_link(
    source_id: String,
    target_id: String,
    relationship_type: String,
    confirm: bool,
    db: &database::Database,
) -> Result<()> {
    // Verify both ideas exist
    let source_idea = db.get_by_id(&source_id).await?;
    let source_idea = match source_idea {
        Some(idea) => idea,
        None => {
            println!("❌ Source idea not found: {}", source_id.red());
            return Ok(());
        }
    };

    let target_idea = db.get_by_id(&target_id).await?;
    let target_idea = match target_idea {
        Some(idea) => idea,
        None => {
            println!("❌ Target idea not found: {}", target_id.red());
            return Ok(());
        }
    };

    let rel_type_enum = match relationship_type.parse::<database::RelationshipType>() {
        Ok(rt) => rt,
        Err(_) => {
            println!(
                "❌ Invalid relationship type: '{}'",
                relationship_type.red()
            );
            println!("Valid types: depends_on, related_to, part_of, parent, child, duplicate, blocks, blocked_by, similar_to");
            return Ok(());
        }
    };

    if confirm {
        println!("🔗 Creating relationship:");
        println!(
            "  Source: {} ({})",
            source_id.blue(),
            source_idea.content.chars().take(50).collect::<String>()
        );
        println!(
            "  Target: {} ({})",
            target_id.blue(),
            target_idea.content.chars().take(50).collect::<String>()
        );
        println!("  Type: {}", relationship_type.cyan());

        let confirmed = Confirm::new()
            .with_prompt("Create this relationship?")
            .interact()?;

        if !confirmed {
            println!("❌ Operation cancelled.");
            return Ok(());
        }
    }

    match db
        .create_relationship(&source_id, &target_id, rel_type_enum)
        .await
    {
        Ok(relationship_id) => {
            println!("✅ Relationship created successfully!");
            println!("   ID: {}", relationship_id.green());
        }
        Err(e) => {
            println!("❌ Failed to create relationship: {}", e.to_string().red());
        }
    }

    Ok(())
}

async fn handle_list_links(idea_id: String, db: &database::Database) -> Result<()> {
    // Verify idea exists
    let idea = db.get_by_id(&idea_id).await?;
    let idea = match idea {
        Some(idea) => idea,
        None => {
            println!("❌ Idea not found: {}", idea_id.red());
            return Ok(());
        }
    };

    let relationships = db.get_relationships_for_idea(&idea_id).await?;

    if relationships.is_empty() {
        println!("📭 No relationships found for idea: {}", idea_id.yellow());
        return Ok(());
    }

    println!(
        "🔗 Relationships for idea: {} ({})",
        idea_id.blue(),
        idea.content.chars().take(50).collect::<String>()
    );
    println!();

    for (i, rel) in relationships.iter().enumerate() {
        // Get the related idea details
        let related_idea_id = if rel.source_idea_id == idea_id {
            &rel.target_idea_id
        } else {
            &rel.source_idea_id
        };

        let related_idea = db.get_by_id(related_idea_id).await?;

        println!(
            "{}. {} {} {}",
            (i + 1).to_string().bright_blue(),
            rel.relationship_type.to_string().cyan(),
            "→".bright_black(),
            related_idea
                .map(|idea| idea.content.chars().take(60).collect::<String>())
                .unwrap_or_else(|| format!("(idea {} not found)", related_idea_id))
                .white()
        );
        println!(
            "   ID: {} ↔ {}",
            rel.source_idea_id.dimmed(),
            rel.target_idea_id.dimmed()
        );
        println!(
            "   Created: {}",
            rel.created_at.format("%Y-%m-%d %H:%M").to_string().dimmed()
        );
        println!();
    }

    Ok(())
}

async fn handle_show_related(
    idea_id: String,
    relationship_type: Option<String>,
    db: &database::Database,
) -> Result<()> {
    // Verify idea exists
    let idea = db.get_by_id(&idea_id).await?;
    let idea = match idea {
        Some(idea) => idea,
        None => {
            println!("❌ Idea not found: {}", idea_id.red());
            return Ok(());
        }
    };

    let rel_type_enum = if let Some(rel_type) = relationship_type {
        match rel_type.parse::<database::RelationshipType>() {
            Ok(rt) => Some(rt),
            Err(_) => {
                println!("❌ Invalid relationship type: '{}'", rel_type.red());
                println!("Valid types: depends_on, related_to, part_of, parent, child, duplicate, blocks, blocked_by, similar_to");
                return Ok(());
            }
        }
    } else {
        None
    };

    let related_ideas = db
        .get_related_ideas(&idea_id, rel_type_enum.clone())
        .await?;

    if related_ideas.is_empty() {
        println!("📭 No related ideas found for: {}", idea_id.yellow());
        return Ok(());
    }

    let filter_text = if let Some(rel_type) = &rel_type_enum {
        format!(" (filtered by: {})", rel_type.to_string().cyan())
    } else {
        "".to_string()
    };

    println!(
        "🔗 Related ideas for: {}{} ({})",
        idea_id.blue(),
        filter_text,
        idea.content.chars().take(50).collect::<String>()
    );
    println!();

    for (i, (related_idea, rel_type)) in related_ideas.iter().enumerate() {
        let score_display = if let Some(score) = related_idea.final_score {
            format!(" 📊 {:.1}/10", score)
        } else {
            "".to_string()
        };

        println!(
            "{}. {} {}",
            (i + 1).to_string().bright_blue(),
            rel_type.to_string().cyan(),
            related_idea.content.white()
        );
        println!(
            "   ID: {}{}",
            related_idea.id.dimmed(),
            score_display.dimmed()
        );
        println!(
            "   Created: {}",
            related_idea
                .created_at
                .format("%Y-%m-%d %H:%M")
                .to_string()
                .dimmed()
        );
        println!();
    }

    Ok(())
}

async fn handle_remove_link(
    relationship_id: String,
    confirm: bool,
    db: &database::Database,
) -> Result<()> {
    // Get details about this relationship
    // Since we don't have a direct getter, we'll just remove by ID

    if confirm {
        println!(
            ".unlink Removing relationship: {}",
            relationship_id.yellow()
        );

        let confirmed = Confirm::new()
            .with_prompt("Permanently remove this relationship?")
            .interact()?;

        if !confirmed {
            println!("❌ Removal cancelled.");
            return Ok(());
        }
    }

    match db.delete_relationship(&relationship_id).await {
        Ok(()) => {
            println!(
                "✅ Relationship removed successfully: {}",
                relationship_id.green()
            );
        }
        Err(e) => {
            println!("❌ Failed to remove relationship: {}", e.to_string().red());
        }
    }

    Ok(())
}

async fn handle_find_path(from: String, to: String, db: &database::Database) -> Result<()> {
    // For now, just a simple implementation that finds direct connections
    // A full pathfinding algorithm would be more complex

    println!("🔍 Finding path from {} to {}...", from.cyan(), to.cyan());

    // Check if there's a direct connection
    let from_idea = db.get_by_id(&from).await?;
    let to_idea = db.get_by_id(&to).await?;

    if from_idea.is_none() {
        println!("❌ Starting idea not found: {}", from.red());
        return Ok(());
    }

    if to_idea.is_none() {
        println!("❌ Target idea not found: {}", to.red());
        return Ok(());
    }

    let from_relationships = db.get_relationships_for_idea(&from).await?;

    // Check if 'to' is a direct relationship of 'from'
    let direct_connection = from_relationships
        .iter()
        .find(|rel| rel.source_idea_id == from || rel.target_idea_id == from);

    if let Some(rel) = direct_connection {
        if rel.source_idea_id == to || rel.target_idea_id == to {
            println!("🎯 Direct connection found!");
            println!(
                "  {} {} {}",
                from.blue(),
                rel.relationship_type.to_string().cyan(),
                to.blue()
            );
            return Ok(());
        }
    }

    // Find indirect paths (simple 2-step search for now)
    let mut paths_found = Vec::new();

    for rel in &from_relationships {
        let connected_idea_id = if rel.source_idea_id == from {
            rel.target_idea_id.clone()
        } else {
            rel.source_idea_id.clone()
        };

        // Check if this connected idea has a relationship with the target
        let connected_relationships = db.get_relationships_for_idea(&connected_idea_id).await?;
        for connected_rel in &connected_relationships {
            let check_id = if connected_rel.source_idea_id == connected_idea_id {
                connected_rel.target_idea_id.clone()
            } else {
                connected_rel.source_idea_id.clone()
            };

            if check_id == to {
                paths_found.push((
                    format!("{} -> {} -> {}", from, connected_idea_id, to),
                    format!(
                        "({} -> {} -> {})",
                        rel.relationship_type, connected_rel.relationship_type, "?"
                    ),
                ));
            }
        }
    }

    if !paths_found.is_empty() {
        println!("🔗 Indirect paths found:");
        for (path, details) in paths_found {
            println!("  {}", path.green());
            println!("  {}", details.dimmed());
        }
    } else {
        println!(
            "❌ No direct or 2-step path found between {} and {}",
            from.red(),
            to.red()
        );
        println!("💡 Try linking ideas that might connect these two concepts");
    }

    Ok(())
}

File: internal/cli/link.go
package cli

import (
	"fmt"

	"github.com/rs/zerolog/log"
	"github.com/ryacub/telos-idea-matrix/internal/cliutil"
	"github.com/ryacub/telos-idea-matrix/internal/models"
	"github.com/spf13/cobra"
)

const (
	ideaNotFoundMessage = "(idea not found)"
)

func newLinkCommand() *cobra.Command {
	cmd := &cobra.Command{
		Use:   "link",
		Short: "Manage relationships between ideas",
		Long: `Manage relationships between ideas.

Link ideas together to express dependencies, hierarchies, and associations.
This helps track how ideas relate to each other and find dependency paths.

Available relationship types:
  - depends_on: Source depends on target
  - related_to: Ideas are related
  - part_of: Source is part of target
  - parent/child: Hierarchical relationship
  - duplicate: Ideas are duplicates
  - blocks/blocked_by: Blocking relationship
  - similar_to: Similar ideas`,
	}

	cmd.AddCommand(newLinkCreateCommand())
	cmd.AddCommand(newLinkListCommand())
	cmd.AddCommand(newLinkShowCommand())
	cmd.AddCommand(newLinkRemoveCommand())
	cmd.AddCommand(newLinkPathCommand())

	return cmd
}

func newLinkCreateCommand() *cobra.Command {
	var noConfirm bool

	cmd := &cobra.Command{
		Use:   "create <source-id> <target-id> <type>",
		Short: "Create a relationship between two ideas",
		Long: `Create a relationship between two ideas.

Relationship types:
  - depends_on: Source depends on target
  - related_to: Ideas are related
  - part_of: Source is part of target
  - parent/child: Hierarchical relationship
  - duplicate: Ideas are duplicates
  - blocks/blocked_by: Blocking relationship
  - similar_to: Similar ideas

Examples:
  tm link create abc123 def456 depends_on
  tm link create abc123 ghi789 related_to --no-confirm`,
		Args: cobra.ExactArgs(3),
		RunE: func(cmd *cobra.Command, args []string) error {
			return runLinkCreate(args[0], args[1], args[2], noConfirm)
		},
	}

	cmd.Flags().BoolVar(&noConfirm, "no-confirm", false, "Skip confirmation prompt")

	return cmd
}

func newLinkListCommand() *cobra.Command {
	return &cobra.Command{
		Use:   "list <idea-id>",
		Short: "List all relationships for an idea",
		Long: `List all relationships for an idea.

Shows both outgoing relationships (where this idea is the source) and
incoming relationships (where this idea is the target).

Examples:
  tm link list abc123`,
		Args: cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return runLinkList(args[0])
		},
	}
}

func newLinkShowCommand() *cobra.Command {
	var relType string

	cmd := &cobra.Command{
		Use:   "show <idea-id>",
		Short: "Show related ideas for an idea",
		Long: `Show related ideas (not relationships) for an idea.

Displays full idea details including content and scores.
Optionally filter by relationship type.

Examples:
  tm link show abc123
  tm link show abc123 --type depends_on`,
		Args: cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return runLinkShow(args[0], relType)
		},
	}

	cmd.Flags().StringVar(&relType, "type", "", "Filter by relationship type")

	return cmd
}

func newLinkRemoveCommand() *cobra.Command {
	var noConfirm bool

	cmd := &cobra.Command{
		Use:   "remove <relationship-id>",
		Short: "Remove a relationship between ideas",
		Long: `Remove a relationship between ideas.

You can get the relationship ID from the 'link list' command.

Examples:
  tm link remove rel123
  tm link remove rel456 --no-confirm`,
		Args: cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return runLinkRemove(args[0], noConfirm)
		},
	}

	cmd.Flags().BoolVar(&noConfirm, "no-confirm", false, "Skip confirmation prompt")

	return cmd
}

func newLinkPathCommand() *cobra.Command {
	var maxDepth int

	cmd := &cobra.Command{
		Use:   "path <source-id> <target-id>",
		Short: "Find dependency paths between ideas",
		Long: `Find all paths between two ideas using breadth-first search.

Useful for understanding how ideas are connected and identifying
dependency chains.

Examples:
  tm link path abc123 xyz789
  tm link path abc123 xyz789 --max-depth 5`,
		Args: cobra.ExactArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			return runLinkPath(args[0], args[1], maxDepth)
		},
	}

	cmd.Flags().IntVar(&maxDepth, "max-depth", 3, "Maximum path length")

	return cmd
}

// --- Implementation Functions ---

func runLinkCreate(sourceID, targetID, relTypeStr string, noConfirm bool) error {
	// Validate relationship type
	relType, err := models.ParseRelationshipType(relTypeStr)
	if err != nil {
		if _, printErr := cliutil.ErrorColor.Printf("❌ Invalid relationship type: %s\n", relTypeStr); printErr != nil {
			log.Warn().Err(printErr).Msg("failed to print error message")
		}
		fmt.Println("\nValid types:")
		for _, rt := range models.AllRelationshipTypes() {
			fmt.Printf("  - %s\n", rt)
		}
		return nil
	}

	// Get both ideas for confirmation
	sourceIdea, err := ctx.Repository.GetByID(sourceID)
	if err != nil {
		if _, printErr := cliutil.ErrorColor.Printf("❌ Source idea not found: %s\n", sourceID); printErr != nil {
			log.Warn().Err(printErr).Msg("failed to print error message")
		}
		return nil
	}

	targetIdea, err := ctx.Repository.GetByID(targetID)
	if err != nil {
		if _, printErr := cliutil.ErrorColor.Printf("❌ Target idea not found: %s\n", targetID); printErr != nil {
			log.Warn().Err(printErr).Msg("failed to print error message")
		}
		return nil
	}

	// Show confirmation
	fmt.Println()
	if _, err := infoColor.Println("Creating relationship:"); err != nil {
		log.Warn().Err(err).Msg("failed to print message")
	}
	fmt.Printf("  Source: [%s] %s\n", truncateID(sourceID), cliutil.TruncateText(sourceIdea.Content, 60))
	fmt.Printf("  Target: [%s] %s\n", truncateID(targetID), cliutil.TruncateText(targetIdea.Content, 60))
	fmt.Printf("  Type: %s\n", relType)
	fmt.Println()

	// Get user confirmation
	if !noConfirm {
		if !cliutil.Confirm("Continue?") {
			if _, err := cliutil.WarningColor.Println("❌ Cancelled."); err != nil {
				log.Warn().Err(err).Msg("failed to print message")
			}
			return nil
		}
	}

	// Create relationship
	relationship, err := models.NewIdeaRelationship(sourceID, targetID, relType)
	if err != nil {
		return fmt.Errorf("failed to create relationship: %w", err)
	}

	err = ctx.Repository.CreateRelationship(relationship)
	if err != nil {
		return fmt.Errorf("failed to save relationship: %w", err)
	}

	if _, err := cliutil.SuccessColor.Printf("✓ Relationship created successfully (ID: %s)\n", truncateID(relationship.ID)); err != nil {
		log.Warn().Err(err).Msg("failed to print message")
	}
	return nil
}

func runLinkList(ideaID string) error {
	// Verify idea exists
	idea, err := ctx.Repository.GetByID(ideaID)
	if err != nil {
		if _, printErr := cliutil.ErrorColor.Printf("❌ Idea not found: %s\n", ideaID); printErr != nil {
			log.Warn().Err(printErr).Msg("failed to print error message")
		}
		return nil
	}

	relationships, err := ctx.Repository.GetRelationshipsForIdea(ideaID)
	if err != nil {
		return fmt.Errorf("failed to get relationships: %w", err)
	}

	if len(relationships) == 0 {
		if _, err := cliutil.WarningColor.Printf("📭 No relationships found for idea: %s\n", truncateID(ideaID)); err != nil {
			log.Warn().Err(err).Msg("failed to print message")
		}
		return nil
	}

	fmt.Println()
	if _, err := infoColor.Printf("🔗 Relationships for idea: %s\n", truncateID(ideaID)); err != nil {
		log.Warn().Err(err).Msg("failed to print message")
	}
	fmt.Printf("   %s\n", cliutil.TruncateText(idea.Content, 60))
	fmt.Println()

	// Separate outgoing and incoming relationships
	var outgoing, incoming []*models.IdeaRelationship
	for _, rel := range relationships {
		if rel.SourceIdeaID == ideaID {
			outgoing = append(outgoing, rel)
		} else {
			incoming = append(incoming, rel)
		}
	}

	// Display outgoing relationships
	if len(outgoing) > 0 {
		if _, err := cliutil.SuccessColor.Println("Outgoing (where this idea is the source):"); err != nil {
			log.Warn().Err(err).Msg("failed to print message")
		}
		for i, rel := range outgoing {
			relatedIdeaID := rel.TargetIdeaID
			relatedIdea, _ := ctx.Repository.GetByID(relatedIdeaID)

			relatedContent := ideaNotFoundMessage
			if relatedIdea != nil {
				relatedContent = cliutil.TruncateText(relatedIdea.Content, 60)
			}

			fmt.Printf("  %d. %s → [%s] %s\n",
				i+1,
				rel.RelationshipType,
				truncateID(relatedIdeaID),
				relatedContent,
			)
			fmt.Printf("     ID: %s | Created: %s\n",
				truncateID(rel.ID),
				rel.CreatedAt.Format("2006-01-02 15:04"),
			)
		}
		fmt.Println()
	}

	// Display incoming relationships
	if len(incoming) > 0 {
		if _, err := infoColor.Println("Incoming (where this idea is the target):"); err != nil {
			log.Warn().Err(err).Msg("failed to print message")
		}
		for i, rel := range incoming {
			relatedIdeaID := rel.SourceIdeaID
			relatedIdea, _ := ctx.Repository.GetByID(relatedIdeaID)

			relatedContent := ideaNotFoundMessage
			if relatedIdea != nil {
				relatedContent = cliutil.TruncateText(relatedIdea.Content, 60)
			}

			fmt.Printf("  %d. %s ← [%s] %s\n",
				i+1,
				rel.RelationshipType,
				truncateID(relatedIdeaID),
				relatedContent,
			)
			fmt.Printf("     ID: %s | Created: %s\n",
				truncateID(rel.ID),
				rel.CreatedAt.Format("2006-01-02 15:04"),
			)
		}
		fmt.Println()
	}

	fmt.Printf("Total: %d relationships\n", len(relationships))
	return nil
}

func runLinkShow(ideaID, relTypeStr string) error {
	// Verify idea exists
	idea, err := ctx.Repository.GetByID(ideaID)
	if err != nil {
		if _, printErr := cliutil.ErrorColor.Printf("❌ Idea not found: %s\n", ideaID); printErr != nil {
			log.Warn().Err(printErr).Msg("failed to print error message")
		}
		return nil
	}

	// Parse relationship type if provided
	var relType *models.RelationshipType
	if relTypeStr != "" {
		rt, err := models.ParseRelationshipType(relTypeStr)
		if err != nil {
			if _, printErr := cliutil.ErrorColor.Printf("❌ Invalid relationship type: %s\n", relTypeStr); printErr != nil {
				log.Warn().Err(printErr).Msg("failed to print error message")
			}
			fmt.Println("\nValid types:")
			for _, t := range models.AllRelationshipTypes() {
				fmt.Printf("  - %s\n", t)
			}
			return nil
		}
		relType = &rt
	}

	relatedIdeas, err := ctx.Repository.GetRelatedIdeas(ideaID, relType)
	if err != nil {
		return fmt.Errorf("failed to get related ideas: %w", err)
	}

	if len(relatedIdeas) == 0 {
		if _, err := cliutil.WarningColor.Printf("📭 No related ideas found for: %s\n", truncateID(ideaID)); err != nil {
			log.Warn().Err(err).Msg("failed to print message")
		}
		return nil
	}

	filterText := ""
	if relType != nil {
		filterText = fmt.Sprintf(" (filtered by: %s)", *relType)
	}

	fmt.Println()
	if _, err := infoColor.Printf("🔗 Related ideas for: %s%s\n", truncateID(ideaID), filterText); err != nil {
		log.Warn().Err(err).Msg("failed to print message")
	}
	fmt.Printf("   %s\n", cliutil.TruncateText(idea.Content, 60))
	fmt.Println()

	for i, relatedIdea := range relatedIdeas {
		scoreDisplay := ""
		if relatedIdea.FinalScore > 0 {
			scoreDisplay = fmt.Sprintf(" 📊 %.1f/10", relatedIdea.FinalScore)
		}

		fmt.Printf("%d. %s\n", i+1, relatedIdea.Content)
		fmt.Printf("   ID: %s | Status: %s%s\n",
			truncateID(relatedIdea.ID),
			relatedIdea.Status,
			scoreDisplay,
		)
		fmt.Printf("   Created: %s\n", relatedIdea.CreatedAt.Format("2006-01-02 15:04"))
		fmt.Println()
	}

	fmt.Printf("Found %d related ideas\n", len(relatedIdeas))
	return nil
}

func runLinkRemove(relationshipID string, noConfirm bool) error {
	// Get relationship details
	rel, err := ctx.Repository.GetRelationship(relationshipID)
	if err != nil {
		if _, printErr := cliutil.ErrorColor.Printf("❌ Relationship not found: %s\n", relationshipID); printErr != nil {
			log.Warn().Err(printErr).Msg("failed to print error message")
		}
		return nil
	}

	// Get idea details for display
	sourceIdea, _ := ctx.Repository.GetByID(rel.SourceIdeaID)
	targetIdea, _ := ctx.Repository.GetByID(rel.TargetIdeaID)

	sourceContent := ideaNotFoundMessage
	if sourceIdea != nil {
		sourceContent = cliutil.TruncateText(sourceIdea.Content, 50)
	}

	targetContent := ideaNotFoundMessage
	if targetIdea != nil {
		targetContent = cliutil.TruncateText(targetIdea.Content, 50)
	}

	fmt.Println()
	if _, err := cliutil.WarningColor.Println("Removing relationship:"); err != nil {
		log.Warn().Err(err).Msg("failed to print message")
	}
	fmt.Printf("  ID: %s\n", truncateID(relationshipID))
	fmt.Printf("  [%s] %s\n", truncateID(rel.SourceIdeaID), sourceContent)
	fmt.Printf("    %s →\n", rel.RelationshipType)
	fmt.Printf("  [%s] %s\n", truncateID(rel.TargetIdeaID), targetContent)
	fmt.Println()

	if !noConfirm {
		if !cliutil.Confirm("Are you sure?") {
			if _, err := cliutil.WarningColor.Println("❌ Removal cancelled."); err != nil {
				log.Warn().Err(err).Msg("failed to print message")
			}
			return nil
		}
	}

	if err := ctx.Repository.DeleteRelationship(relationshipID); err != nil {
		return fmt.Errorf("failed to remove relationship: %w", err)
	}

	if _, err := cliutil.SuccessColor.Printf("✓ Relationship removed successfully\n"); err != nil {
		log.Warn().Err(err).Msg("failed to print message")
	}
	return nil
}

func runLinkPath(sourceID, targetID string, maxDepth int) error {
	// Verify both ideas exist
	sourceIdea, err := ctx.Repository.GetByID(sourceID)
	if err != nil {
		if _, printErr := cliutil.ErrorColor.Printf("❌ Source idea not found: %s\n", sourceID); printErr != nil {
			log.Warn().Err(printErr).Msg("failed to print error message")
		}
		return nil
	}

	_, err = ctx.Repository.GetByID(targetID)
	if err != nil {
		if _, printErr := cliutil.ErrorColor.Printf("❌ Target idea not found: %s\n", targetID); printErr != nil {
			log.Warn().Err(printErr).Msg("failed to print error message")
		}
		return nil
	}

	fmt.Println()
	if _, err := infoColor.Printf("🔍 Finding paths from %s to %s...\n", truncateID(sourceID), truncateID(targetID)); err != nil {
		log.Warn().Err(err).Msg("failed to print message")
	}
	fmt.Println()

	paths, err := ctx.Repository.FindRelationshipPath(sourceID, targetID, maxDepth)
	if err != nil {
		return fmt.Errorf("failed to find path: %w", err)
	}

	if len(paths) == 0 {
		if _, err := cliutil.WarningColor.Printf("❌ No path found between %s and %s\n", truncateID(sourceID), truncateID(targetID)); err != nil {
			log.Warn().Err(err).Msg("failed to print message")
		}
		fmt.Println()
		fmt.Println("💡 Try linking ideas that might connect these two concepts")
		return nil
	}

	// Display all found paths
	for i, path := range paths {
		if _, err := cliutil.SuccessColor.Printf("Path %d (%d hops):\n", i+1, len(path)); err != nil {
			log.Warn().Err(err).Msg("failed to print message")
		}

		// Start with source idea
		fmt.Printf("  [%s] %s\n", truncateID(sourceID), cliutil.TruncateText(sourceIdea.Content, 50))

		// Display each step in the path
		currentID := sourceID
		for _, rel := range path {
			// Determine next ID
			var nextID string
			if rel.SourceIdeaID == currentID {
				nextID = rel.TargetIdeaID
			} else {
				nextID = rel.SourceIdeaID
			}

			// Get next idea
			nextIdea, _ := ctx.Repository.GetByID(nextID)
			nextContent := ideaNotFoundMessage
			if nextIdea != nil {
				nextContent = cliutil.TruncateText(nextIdea.Content, 50)
			}

			fmt.Printf("    → %s →\n", rel.RelationshipType)
			fmt.Printf("  [%s] %s\n", truncateID(nextID), nextContent)

			currentID = nextID
		}
		fmt.Println()
	}

	fmt.Printf("Found %d path(s)\n", len(paths))
	return nil
}

// --- Helper Functions ---

func truncateID(id string) string {
	if len(id) <= 8 {
		return id
	}
	return id[:8]
}
